src/extension/vscode/client.ts

```diff
@@ -0,0 +1,162 @@
+  url: string | URL
+): {
+  connected: Promise<void>;
+  unregister: (reason?: Reason) => void;
+  onCleanup: (cleanup: (reason: Reason) => void) => void;
+} {
+  const { signal, cleanup, registerCleanup } = getCleanupController();
+
+  function getClient() {
+    const client = clientRef.deref();
+    if (client) {
+      return getPrivateAccess(client);
+    } else {
+      cleanup("CLIENT_GC");
+    }
+  }
+
+  const ws = new WebSocket(url);
+  registerCleanup(ws.close.bind(ws, 1000));
+  ws.addEventListener("close", cleanup.bind(undefined, "WS_DISCONNECTED"), {
+    once: true,
+    signal,
+  });
+  ws.addEventListener("error", makeErrorHandler(cleanup), {
+    once: true,
+    signal,
+  });
+
+  const wsAdapter = createWsAdapter(ws, signal);
+  const wsRpcClient = createRpcClient(wsAdapter);
+  const wsRpcHandler = createRpcHandler(wsAdapter);
+  const wsActor = createActor(wsAdapter);
+  function getQueriesForClient() {
+    return getQueries(
+      getClient()?.queryManager.getObservableQueries("active") ?? new Map()
+    );
+  }
+  function getMutationsForClient() {
+    return getMutations(getClient()?.queryManager.mutationStore ?? {});
+  }
+  registerCleanup(
+    wsRpcHandler("getQueries", getQueriesForClient),
+    wsRpcHandler("getMutations", getMutationsForClient),
+    wsRpcHandler("getCache", () => getClient()?.cache.extract(true) ?? {}),
+    handleExplorerRequests(wsActor, getClient)
+  );
+
+  ws.addEventListener(
+    "open",
+    function open() {
+      const client = getClient();
+      if (!client) {
+        return;
+      }
+      wsActor.send({
+        type: "registerClient",
+        payload: {
+          id: "",
+          name:
+            "devtoolsConfig" in client ? client.devtoolsConfig.name : undefined,
+          version: client.version,
+          queryCount: getQueriesForClient().length,
+          mutationCount: getMutationsForClient().length,
+        },
+      });
+      loadErrorCodes(wsRpcClient, client.version);
+    },
+    { once: true, signal }
+  );
+
+  return {
+    connected: new Promise<void>((resolve) =>
+      ws.addEventListener("open", () => resolve(), { once: true, signal })
+    ),
+    unregister: cleanup,
+    onCleanup: registerCleanup,
+  };
+}
+
+function createWsAdapter(ws: WebSocket, signal: AbortSignal): MessageAdapter {
+  return {
+    addListener(listener) {
+      const wrappedListener = (ev: MessageEvent<any>) => {
+        listener(JSON.parse(ev.data));
+      };
+      ws.addEventListener("message", wrappedListener, { signal });
+      return () => ws.removeEventListener("message", wrappedListener);
+    },
+    postMessage(message) {
+      ws.send(JSON.stringify(message));
+    },
+  };
+}
+
+function getCleanupController() {
+  const cleanupContoller = new AbortController();
+  const cleanup = cleanupContoller.abort.bind(cleanupContoller);
+  const signal = cleanupContoller.signal;
+  setMaxListeners(20, signal);
+  function registerCleanup(
```

Comment:
```suggestion
  function onCleanup(
```

Maybe just me, but I kinda like calling this `on*` since this is basically just storing event listeners. Thoughts?

Reply:
done